dash/_patch.py

```diff
@@ -0,0 +1,117 @@
+            self._location = location
+        else:
+            # pylint: disable=consider-using-ternary
+            self._location = (parent and parent._location) or []
+        if parent is not None:
+            self._operations = parent._operations
+        else:
+            self._operations = []
+
+    def __getitem__(self, item):
+        return Patch(location=self._location + [item], parent=self)
+
+    def __getattr__(self, item):
+        if item == "tolist":
+            # to_json fix
+            raise AttributeError
+        if item == "_location":
+            return self._location
+        if item == "_operations":
+            return self._operations
+        return self.__getitem__(item)
+
+    def __setattr__(self, key, value):
+        if key in ("_location", "_operations"):
+            self.__dict__[key] = value
+        else:
+            self.__setitem__(key, value)
+
+    def __delattr__(self, item):
+        self.__delitem__(item)
+
+    def __setitem__(self, key, value):
+        if value is _noop:
+            # The += set themselves.
+            return
+        self._operations.append(
+            _operation(
+                "Assign",
+                self._location + [key],
+                value=value,
+            )
+        )
+
+    def __delitem__(self, key):
+        self._operations.append(_operation("Delete", self._location + [key]))
+
+    def __add__(self, other):
+        self._operations.append(_operation("Add", self._location, value=other))
+        return _noop
+
+    def __iadd__(self, other):
+        self._operations.append(_operation("Add", self._location, value=other))
+        return _noop
+
+    def __sub__(self, other):
+        self._operations.append(_operation("Sub", self._location, value=other))
+        return _noop
+
+    def __isub__(self, other):
+        self._operations.append(_operation("Sub", self._location, value=other))
+        return _noop
+
+    def __mul__(self, other):
+        self._operations.append(_operation("Mul", self._location, value=other))
+        return _noop
+
+    def __imul__(self, other):
+        self._operations.append(_operation("Mul", self._location, value=other))
+        return _noop
+
+    def __truediv__(self, other):
+        self._operations.append(_operation("Div", self._location, value=other))
+        return _noop
+
+    def __itruediv__(self, other):
+        self._operations.append(_operation("Div", self._location, value=other))
+        return _noop
+
+    def append(self, item):
+        self._operations.append(_operation("Append", self._location, value=item))
+
+    def prepend(self, item):
+        self._operations.append(_operation("Prepend", self._location, value=item))
+
+    def extend(self, item):
+        if not isinstance(item, list):
+            raise TypeError(f"{item} should be a list")
+        self._operations.append(_operation("Extend", self._location, value=item))
+
+    def merge(self, item):
```

Comment:
`merge` isn't a built-in dict method, but `update` is, can we switch to `update` and mimic its behavior? What you have is the beginning of `update` but it can do more, mainly kwargs.

```
update(...) method of builtins.dict instance
    D.update([E, ]**F) -> None.  Update D from dict/iterable E and F.
    If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k]
    If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v
    In either case, this is followed by: for k in F:  D[k] = F[k]
```

Dicts have `clear` and `pop` methods that, like lists, I don't think we need to implement but could benefit from the same stub I suggested for lists. Other than that the only one I see we could potentially implement is `__ior__`, which is equivalent to `update` ie `p |= {"new": "items"}`

Reply:
I think `pop` don't make sense since we can't return the value.
`clear` can be emulated with ramda empty: https://ramdajs.com/docs/#empty so let's just add it.

Reply:
Oh nice, `empty` works like `clear` for both lists and dicts - let's do it! Agreed re: `pop`.